packages/hydrogen/src/cache/create-with-cache.ts

```diff
@@ -18,45 +20,82 @@ type CreateWithCacheOptions = {
- * like `fetch` and caches the result according to the strategy provided.
- * Use this to call any third-party APIs from loaders or actions.
- *
- * > Note:
- * > Sometimes a request to a third-party API might fail, so you shouldn't cache the result.
- * > To prevent caching, throw when a request fails. If you don't throw, then the result is cached.
+ * Creates utility functions to store data in cache with stale-while-revalidate support.
+ * - Use `withCache.fetch` to simply fetch data from a third-party API.
+ * - Use the more advanced `withCache.run` to execute any asynchronous operation.
  */
 export function createWithCache<T = unknown>(
   cacheOptions: CreateWithCacheOptions,
-): CreateWithCacheReturn<T> {
+) {
   const {cache, waitUntil, request} = cacheOptions;
 
-  return function withCache<T = unknown>(
-    cacheKey: CacheKey,
-    strategy: CachingStrategy,
-    actionFn: ({addDebugData}: CacheActionFunctionParam) => T | Promise<T>,
-  ) {
-    return runWithCache<T>(cacheKey, actionFn, {
-      strategy,
-      cacheInstance: cache,
-      waitUntil,
-      debugInfo: {
-        ...getDebugHeaders(request),
-        stackInfo: getCallerStackLine?.(),
-      },
-    });
+  return {
+    /**
+     * Utility function that executes asynchronous operations and caches the
+     * result according to the strategy provided. Use this to do any type
+     * of asynchronous operation where `withCache.fetch` is insufficient.
+     * For example, when making multiple calls to a third-party API where the
+     * result of all of them needs to be cached under the same cache key.
+     * Whatever data is returned from the `actionFn` will be cached according
+     * to the strategy provided.
+     * Use the `CachingStrategy` to define a custom caching mechanism for your data.
+     * Or use one of the built-in caching strategies: `CacheNone`, `CacheShort`, `CacheLong`.
+     * > Note:
+     * > To prevent caching the result you must throw an error. Otherwise, the result will be cached.
+     * > For example, if you call `fetch` but the response is not successful (e.g. status code >= 400),
+     * > you should throw an error. Otherwise, the response will be cached.
+     */
+    run<InferredActionReturn = T>(
+      cacheKey: CacheKey,
+      strategy: CachingStrategy,
+      actionFn: ({
+        addDebugData,
+      }: CacheActionFunctionParam) =>
+        | InferredActionReturn
+        | Promise<InferredActionReturn>,
+    ) {
+      return runWithCache(cacheKey, actionFn, {
+        strategy,
+        cacheInstance: cache,
+        waitUntil,
+        debugInfo: {
+          ...getDebugHeaders(request),
+          stackInfo: getCallerStackLine?.(),
+        },
+      });
+    },
+
+    /**
+     * Fetches data from a URL and caches the result according to the strategy provided.
+     * When the response is not successful (e.g. status code >= 400), the caching is
+     * skipped automatically and the returned `data` is `null`.
```

Comment:
>I just think that there sometimes might be logic even when something fails

Yes yes, I agree. I guess we are just discussing *how* we get that payload after errors.

>Sometimes it won't even be a string if the 500 has no body

In this case we can return empty string, which is not incorrect I guess (if we go with `data: string | MyData`)

>Perhaps we just mark it as unknown?

In this case, `data: unknown | MyData` turns always into `unknown`.

The only way to make it work would be using a different property to return the error body (as mentioned [here](https://github.com/Shopify/hydrogen/pull/2546#discussion_r1766072096)) or by returning a top level Response-like payload (mentioned [here](https://github.com/Shopify/hydrogen/pull/2546#discussion_r1766183493)).

Which one would you like more?

cc @wizardlyhel thoughts?

Reply:
Since we are allowing devs to pass in their own type, wouldn't they know what to expect the shape of the data even if the response is not ok.

So technically, in our code, we just need `data: T` and devs should supply `cache.fetch<MyData | MyErrorData | null>(...)` to anticipate the shape of the returned data, even when it errors.

Reply:
It's pretty common for the error type to be different than the 200 type.

Reply:
would it make sense to return `response.clone().body`?